Make queries thread safe

Zoxc opened the pull request:
This makes queries thread safe by removing the query stack and making queries point to their parents. Queries write to the query map when starting and cycles are detected by checking if there's already an entry in the query map. This makes cycle detection O(1) instead of O(n), where `n` is the size of the query stack.

This is mostly corresponds to the method I described [here](https://internals.rust-lang.org/t/parallelizing-rustc-using-rayon/6606).

cc @rust-lang/compiler

r? @michaelwoerister

michaelwoerister commented on src/librustc/ty/maps/plumbing.rs:
```diff
@@ -277,10 +252,43 @@ macro_rules! define_maps {
+                    let job = if let Some(job) = job {
+                        job
+                    } else {
+                        break lock;
```
Is there a reason why this line is not just `return Ok(lock);`? I think that would be clearer.

Zoxc replied:
Yeah, that would be better. The `break` is there because this used to be part of another function.

Zoxc replied:
I've fixed this.

michaelwoerister replied:
:+1:

pietroalbini commented:
Ping from triage @michaelwoerister! The PR author pushed new commits, could you review them?

michaelwoerister commented:
I haven't forgotten about this, I'm just busy with bug fixing the last few days.

nikomatsakis commented:
Same here. I have some time scheduled for review later today, though I expect i'm largely going to defer to @michaelwoerister and/or @eddyb here since I don't have a ton of time right now.

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1635,21 +1651,34 @@ impl<'a, 'tcx> Lift<'tcx> for &'a Slice<Predicate<'a>> {
+    use rustc_data_structures::sync::Lrc;
+
+    #[derive(Clone)]
+    pub struct ImplicitCtxt<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
```
Needs documentation. What is an `ImplicitCtxt`?

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1484,11 +1486,25 @@ impl<'a, 'tcx> TyCtxt<'a, 'tcx, 'tcx> {
 impl<'gcx: 'tcx, 'tcx> GlobalCtxt<'gcx> {
     /// Call the closure with a local `TyCtxt` using the given arena.
-    pub fn enter_local<F, R>(&self, arena: &'tcx DroplessArena, f: F) -> R
+    pub fn enter_local<F, R>(&self,
```
If we're going to reformat, let's adopt the modern style:

```
pub fn enter_local<F, R>(
    &self,
    arena: &'tcx DroplessArena,
    f: F,
) -> R
where
    F: ...
{
```

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        })
+    }
+
+    pub fn with_thread_locals<F, R>(f: F) -> R
```
Comment please. (I know... the original code isn't that documented either...c'est la vie)

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        })
+    }
+
+    pub fn enter_context<'a, 'gcx: 'tcx, 'tcx, F, R>(context: &ImplicitCtxt<'a, 'gcx, 'tcx>,
```
Nit: as above, let's adopt the modern style

nikomatsakis replied:
Also, comment =) 

One thing I am wondering about: why does this take an `&ImplicitCtxt` versus creating one locally?

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        }
+    }
+
+    pub fn with_fully_related_context<'a, 'gcx, 'tcx, F, R>(tcx: TyCtxt<'a, 'gcx, 'tcx>, f: F) -> R
```
Comment: what is a "fully related" context?

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        })
+    }
+
+    pub fn with_context_opt<F, R>(f: F) -> R
```
Comment: what does this do? I guess it gets the context from the environment?

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
         })
     }
 
+    pub fn with_related_context<'a, 'gcx, 'tcx1, F, R>(tcx: TyCtxt<'a, 'gcx, 'tcx1>, f: F) -> R
```
Comment: what is a "related context"?

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        })
+    }
+
+    pub fn with_context<F, R>(f: F) -> R
```
Comment: What is a "context" and how is that different from a "related" context =)

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        with_context_opt(|opt_context| f(opt_context.expect("no ImplicitCtxt stored in tls")))
+    }
+
     pub fn with<F, R>(f: F) -> R
```
Pre-existing, but I'm on a roll: let's comment this too...

nikomatsakis commented on src/librustc/ty/context.rs:
```diff
@@ -1658,59 +1687,120 @@ pub mod tls {
+        with_context(|context| f(context.tcx))
     }
 
     pub fn with_opt<F, R>(f: F) -> R
```
Likewise

nikomatsakis commented on src/librustc/ty/maps/job.rs:
```diff
@@ -0,0 +1,90 @@
+// except according to those terms.
+
+#![allow(warnings)]
+
```
Ideally this would have a module comment pointing into the rustc-guide chapter that describes the design =)

nikomatsakis replied:
Basically every struct and public function in this module should have some form of doc comment

nikomatsakis commented on src/librustc/ty/maps/job.rs:
```diff
@@ -0,0 +1,90 @@
+}
+
+impl<'tcx> QueryJob<'tcx> {
+    pub fn new(entry: StackEntry<'tcx>,
```
Can you please run rustfmt on this file? <3

nikomatsakis commented on src/librustc/ty/maps/mod.rs:
```diff
@@ -389,7 +393,7 @@ define_maps! { <'tcx>
         CanonicalProjectionGoal<'tcx>
     ) -> Result<
-        Lrc<Canonical<'tcx, QueryResult<'tcx, NormalizationResult<'tcx>>>>,
+        Lrc<Canonical<'tcx, canonical::QueryResult<'tcx, NormalizationResult<'tcx>>>>,
```
Hmm. We should probably rename `canonical::QueryResult`, but not in this PR.

nikomatsakis commented on src/librustc/ty/maps/plumbing.rs:
```diff
@@ -277,10 +252,47 @@ macro_rules! define_maps {
+            fn try_get_lock(tcx: TyCtxt<'a, $tcx, 'lcx>,
+                            mut span: Span,
+                            key: &$K)
+                            -> Result<LockGuard<'a, QueryMap<$tcx, Self>>,
```
Can we make a custom enum here? This type is sort of incomprehensible. Something with named variants seems better. Like:

```rust
enum TryGetLock {
    NotYetStarted(Guard), // Today: Ok(guard)
    JobCompleted(...), // Today: Err(Ok(...))
    CycleDetected(...), // Today: Err(Err...))
}
```

I guess it means:

- `Ok(guard)` -- no job yet started, lock is still held
- `Err(Ok(...))` -- cache hit, or else we blocked on a job and it completed
- `Err(Err(..))` -- cycle detected

nikomatsakis commented on src/librustc/ty/maps/plumbing.rs:
```diff
@@ -294,24 +306,39 @@ macro_rules! define_maps {
-                    profq_msg!(tcx, ProfileQueriesMsg::CacheHit);
-                    tcx.dep_graph.read_index(value.index);
-                    return Ok((&value.value).clone());
+                macro_rules! get_lock {
```
why is this a macro and not a function?

nikomatsakis replied:
Oh, because it returns I guess? Hmm. OK. Maybe we can rename it `get_lock_or_return` or something.

nikomatsakis commented on src/librustc/ty/maps/plumbing.rs:
```diff
@@ -391,6 +421,63 @@ macro_rules! define_maps {
+                            // Also signal the completion of the job, so waiters
+                            // will continue execution
+                            job.signal_complete();
+                        });
```
Neat.

Zoxc commented:
@bors try

Zoxc commented:
@bors try

bors commented:
:hourglass: Trying commit 31c0581aa0a405f94d031f8570ba2f2938ccf2d1 with merge c83390e645c0ae1d55d6f56636824a76fe084663...

bors commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/rust-lang/rust/builds/356368689?utm_source=github_status&utm_medium=notification)
State: approved= try=True

Zoxc commented:
@Mark-Simulacrum Can I get a perf run?

Mark-Simulacrum commented:
Perf queued.

Mark-Simulacrum commented:
http://perf.rust-lang.org/compare.html?start=c19264fa835a1eca86de4fd2e86a87b3919e57cf&end=c83390e645c0ae1d55d6f56636824a76fe084663&stat=instructions%3Au

Zoxc commented:
I've renamed `StackEntry` to `QueryInfo` since there's no longer a stack. I removed the `track_diagnostics` field of `QueryJob` since it was no longer used. I also added some comments.

michaelwoerister commented:
Sorry for not getting to this yet, @Zoxc. Either we'll take a look at this together during the work week or I'll do an in-depth review on Tuesday after Easter.

michaelwoerister commented:
OK, I finally had time to review this. The comments you added since the initial version make things a lot clearer, thanks, @Zoxc! I think we can go ahead an merge this. The basic strategy seems sound to me and a good starting point. 

@nikomatsakis, are you satisfied with the comments added or do you want more? Otherwise, r=me.

michaelwoerister commented:
I'm going to go ahead and r+ this for now. We can always add more comments later.

@bors r+

bors commented:
:pushpin: Commit 4f7d0fd has been approved by `michaelwoerister`

<!-- @bors r=michaelwoerister 4f7d0fde1c5f577c1f956d5d4edfbb202a5bc3cf -->

bors commented:
:hourglass: Testing commit 4f7d0fde1c5f577c1f956d5d4edfbb202a5bc3cf with merge 7222241e7c2d7caf9ad6ee6e34748e4addfb8dd3...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6922), [status-travis](https://travis-ci.org/rust-lang/rust/builds/362718000?utm_source=github_status&utm_medium=notification)
Approved by: michaelwoerister
Pushing 7222241e7c2d7caf9ad6ee6e34748e4addfb8dd3 to master...

michaelwoerister commented:
:tada:
